metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_classes.py

```diff
@@ -0,0 +1,170 @@
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, List, Optional
+
+
+@dataclass
+class Workspace:
+    """
+    PowerBi Workspace
+    """
+
+    id: str
+    name: str
+    dashboards: List["Dashboard"]
+    reports: List["Report"]
+    datasets: Dict[str, "PowerBIDataset"]
+    report_endorsements: Dict[str, List[str]]
+    dashboard_endorsements: Dict[str, List[str]]
+
+
+@dataclass
+class DataSource:
+    """
+    PowerBi
+    """
+
+    id: str
+    type: str
+    raw_connection_detail: Dict
+
+    def __members(self):
+        return (self.id,)
+
+    def __eq__(self, instance):
+        return (
+            isinstance(instance, DataSource)
+            and self.__members() == instance.__members()
+        )
+
+    def __hash__(self):
+        return hash(self.__members())
```

Comment:
I know this code has been there for a while, but it seems really worrying - why do these classes need eq and hash?

the users of these objects should dedup based on the id directly instead of this

Reply:
built-in set required this to be implemented

Reply:
in that case, we should be using a dict of id -> actual object for deduplication

there's almost never a good reason to be overriding __hash__, especially for a mutable type